### PR TITLE
[numpy] torch.{all, any} : Extend Dtype Support

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #44779

### CODE REVIEW

**Review by heitorschueroff — APPROVED** [review]
LGTM. Thanks for this PR, it's a very welcomed change.

Note: Now that torch.all and torch.any supports all dtypes, we should document it in the public APIs as mentioned here https://github.com/pytorch/pytorch/issues/44779, but this can be a separate PR.

**Review by heitorschueroff — CHANGES_REQUESTED** [review]
It looks like the XLA failures are related to the changes. Could you look into what's causing it please?

**Comment by kshitij12345** [comment]
@heitorschueroff Thanks for looking at it. 

As for XLA, I m not really sure what is happening. 
Maybe @JackCaoG can have a look. 

Thanks!

**Comment by JackCaoG** [comment]
@kshitij12345 XLA change is ready, I will merge it when this pr is merged.

**Comment by kshitij12345** [comment]
@heitorschueroff Have fixed the conflict. ROCm failure looks irrelevant.

**Comment by heitorschueroff** [comment]
> @heitorschueroff Have fixed the conflict. ROCm failure looks irrelevant.

Thank you for this contribution, I'm important your changes now.

**Review by heitorschueroff — APPROVED** [review]

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:235]: include_bool -> includeBool

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:266]: include_bool -> includeBool

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:238]: Avoid the implicit cast with:
`[=](uint8_t a, uint8_t b) -> uint8_t { return ((a && b) ? 1 : 0); },`

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:269]: Avoid the implicit cast with:
[=](uint8_t a, uint8_t b) -> uint8_t { return ((a && b) ? 1 : 0); },

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:252]: Avoid implicit cast with:
`c[i] = ((a[i] && b[i]) ? 1 : 0);`

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:273]: Avoid implicit cast with:
`c[i] = ((a[i] && b[i]) ? 1 : 0);`

**Review by heitorschueroff — CHANGES_REQUESTED** [review]
It looks like I missed some details in my review. Our internal tests on phabricator are complaining. I left some comments from phabricator, they should be fairly quick to fix and then I can land it without problems.

**Comment by kshitij12345** [comment]
@heitorschueroff Done.

**Review by heitorschueroff — APPROVED** [review]

**Comment by kshitij12345** [comment]
@JackCaoG Please merge the XLA fix. Thanks!

**Comment by ngimel** [comment]
@kshitij12345 there are a few problems with this PR
1) as noted already, it breaks XLA CI, @heitorschueroff please avoid landing things before XLA fix is merged
2) it produces results of the same dtype as input, this is different from numpy:
```
In [9]: a=torch.randn(3,3)                                                                                                                                                                    
In [10]: anp=a.numpy()                                                                                                                                                                        
In [11]: np.all(anp, 0)                                                                                                                                                                       
Out[11]: array([ True,  True,  True])
In [12]: a.all(0)                                                                                                                                                                             
Out[12]: tensor([1., 1., 1.])
```
3) it introduces performance cliff for non-bool input types, because it does not use binary_kernel_reduce_vec, and instead uses binary_kernel_reduce which is much (5-10x) slower.

Can you please work on fixing 2) and 3) ?

**Comment by JackCaoG** [comment]
> @kshitij12345 there are a few problems with this PR
> 
> 1. as noted already, it breaks XLA CI, @heitorschueroff please avoid landing things before XLA fix is merged
> 2. it produces results of the same dtype as input, this is different from numpy:
> 
> ```
> In [9]: a=torch.randn(3,3)                                                                                                                                                                    
> In [10]: anp=a.numpy()                                                                                                                                                                        
> In [11]: np.all(anp, 0)                                                                                                                                                                       
> Out[11]: array([ True,  True,  True])
> In [12]: a.all(0)                                                                                                                                                                             
> Out[12]: tensor([1., 1., 1.])
> ```
> 
> 1. it introduces performance cliff for non-bool input types, because it does not use binary_kernel_reduce_vec, and instead uses binary_kernel_reduce which is much (5-10x) slower.
> 
> Can you please work on fixing 2) and 3) ?

I merged XLA change. If this pr will be reverted I will revert the xla pr as well. Otherwise I will work on a companion pr to fix the result type.

**Comment by gchanan** [comment]
I'd look at the `logical_or` and `logical_and` kernels for how to get the dtypes correct; note that it's a little different because of the performance issue @ngimel mentioned above.

**Comment by ngimel** [comment]
Also please don't forget to update documentation.

**Comment by kshitij12345** [comment]
@ngimel 
For 2, it would be a BC-Breaking to change the result-type to `bool`,
I think it would be better to add a warning first otherwise it may break code for users. Let me know what you think about that.

Behaviour in previous version (1.5.1)
```python
>>> import torch
>>> torch.__version__
'1.5.1+cu101'
>>> x = torch.zeros(3,3)
>>> x.to(torch.uint8).all()
tensor(0, dtype=torch.uint8)
>>> x.to(torch.bool).all()
tensor(False)
```

**Comment by ngimel** [comment]
cc @mruberry for deprecating return type for uint8. In any case, for all other types there are no bc breaking concerns, so we should implement correct behavior.

**Comment by mruberry** [comment]
I would try to update the uint8 behavior to be consistent and document the change as BC-breaking. If a scripted network relies on the current behavior (extremely unlikely) we can write an upgrader.

**Pull request closed without merging**
